every symbol becomes 1st class; defines 0-cost lambda and aliases; generics/iterators/templates/etc can be passed to any routine

timotheecour opened the pull request:
This PR makes every symbol 1st class, which enables these:
* passing any symbol in scope (generic/proc/macro/template/iterator/module/etc) to any routine (proc/iterator/macro/etc). This works even if the symbol is overloaded, or if it's a template/macro with all optional params
* aliases (`const myecho = alias echo`) that work with any symbol; can also return a tuple, eg `(fun1: alias echo, fun2: alias doAssert)`
* 0-cost lambdas (eg: `x~>x mod 2 == 0`, `(x,y) ~> x*y`, `lambdaIt: it*2` etc), which are defined in library code on top of `alias`
* composable iterators, which allows define lazy functional programming primitives (eager `toSeq` and lazy `map/filter/join` etc; which can have speed benefit compared to `sequtils`); the resulting code is quite nice and compact, see `tlambda_iterators.nim` which uses library `lambdaIter` that builds on top of `lambda`; it allows defining primitives that work regardless we're passing a value (`@[1,2,3]`) or a lazy iterate (eg `iota(3)`)

Extensive unittests are included, see `tlambda.nim`, `tlambda_iterators.nim`

## fixed issues
* https://github.com/nim-lang/Nim/issues/11870 symbols of no argument templates can't be passed to macros
  see tests: every symbol can be passed to every routine, eg `foo(alias mysym)`
* https://github.com/nim-lang/Nim/issues/4516 Cannot pass inline iterator to inline iterator
  see tests
* https://github.com/nim-lang/Nim/issues/7435 type inference with lambdas doesn't work (requires explicit types)
  see tests, eg: `myfun(10, a~>a*2)`
* fix https://github.com/nim-lang/Nim/issues/5702 (see unittests; can pass builtin `+` just like any other symbol)
* supersedes https://github.com/nim-lang/Nim/pull/8679 (lambdas); current PR has none of the drawbacks of nim-lang/Nim#8679:
  - in particular lambdas are well typed instead of having to be passed as `untyped`
  - usage is simpler since there's no need for an extra `makeLambda` inside each user of the lambda
  - the lambda can be passed as a generic param in `proc callFun[Fun](fun: Fun)`: `callFun(a~>a*10)` or `callFun(myTemplate)` etc, requiring 0 change in existing code (since the lambda can be passed as a generic)
* https://github.com/nim-lang/Nim/issues/8675 RFC: Lambda expression (with inferred placeholders)
* can be used to fix https://github.com/nim-lang/Nim/pull/8325 (see tests that implement show map, filter, join, toSeq, etc)
* implements https://github.com/nim-lang/Nim/issues/9374 ([RFC] first class inline iterators: `iter it = myiter()` to allow perfect forwarding of iterators, with delayed evaluation): see `lambdaIter` in unittests
* can be used to fix https://github.com/nim-lang/RFCs/issues/512 because `toSeq` could be rewritten to be without `untyped` as argument to accept iterators, see `tlambda_iterators.nim` ; likewise, https://github.com/nim-lang/Nim/issues/9219 could be closed since we can now pass the iterator as done in unittests
* closes nim-lang/Nim#8935 since we can do `const testForMe=alias assert`, ditto with aliasing generic/template/macro/overloads 
* closes nim-lang/Nim#11731 since `const my_echo=alias system.echo` works
* using alias instead of static solves this https://github.com/nim-lang/Nim/issues/9679 (Generic types as static parameters: type mismatch), so might be closable too
* would fix https://github.com/nim-lang/Nim/issues/8303 (see map defined in `tlambda_iterators.nim`)
* would fix https://github.com/nim-lang/Nim/issues/16439
* fixes nim-lang/Nim#7090 (alias declarations) ; this was wrongly closed: the workarounds didn't work in many cases and each case required a different syntax); this feature request keeps appearing, eg:
  - https://forum.nim-lang.org/t/5015  
  - https://forum.nim-lang.org/t/1515 
  - btbytes/nim-cookbook#21
  - https://www.reddit.com/r/nim/comments/5v2rgw/alias_echo_to_print/ 
  - https://forum.nim-lang.org/t/1810 

## speed benefit
* can be used to improve speed and ease of use of procs that have callback (functor) parameters: the callback can now be a zero-cost lambda instead of a closure (which incurs overhead, eg due to lack of inlining once we have a closure);

for example the following code gives a **3X to 4X speedup** (with both debug build and `-d:danger`) over algorithms.isSorted which uses a callback closure via
`func isSorted*[T](a: openArray[T], cmp: proc(x, y: T): int {.closure.}, order = SortOrder.Ascending): bool =`

```nim
func isSorted2*[T, Fun](a: openArray[T], cmp: Fun): bool =
  result = true
  for i in 0..<len(a)-1:
    if not cmp(a[i],a[i+1]):
      return false

# benchmark code:
  let n = 1_000
  let m = 100000
  var s = newSeq[int](n)
  for i in 0..<n: s[i] = i*2
  benchmarkDisp("isSorted2", m): doAssert isSorted2(s, (a,b)~>a<b)
  benchmarkDisp("isSorted", m): doAssert isSorted(s, cmp)
```

## other benefits
* with this PR, patterns like `macro mydefine(name: untyped, val: typed)` can now become: `const myname = mydefineAux(val)` ; see tests
* supersedes https://github.com/nim-lang/Nim/pull/11822 (`alias: myecho=echo`)
* partially addresses https://github.com/nim-lang/RFCs/issues/1, see tlambda_iterators.nim
* reduces need for passing a symbol as `untyped` followed by calling `bindSym` since we can now pass the symbol directly (either via an explicit generic param `foo[T](fun: T)` or implicit generic `foo(fun: aliassym)`)
* helps with things like https://nimble.directory/pkg/zerofunctional (the resulting iterator can now be returned with this PR)
* this actually **generalizes `typedesc` and `static`** without their restrictions (see nim-lang/Nim#9679) ; using `alias` you can also pass a type or a const (via passing a template symbol that returns that const, eg `myConstReturningTemplate` or `()~>myConstVal`); maybe `static/typedesc` could now be implemented in terms of `alias` which would fix bugs inherent with these
* provides a cleaner alternative to using `mixin`, for example this helps with issues like https://github.com/nim-lang/Nim/issues/11225 (Generics "sandwiched" between two modules don't mixin their scope symbols properly)

## notes
* https://github.com/nim-lang/Nim/issues/8759 is a minor bug that affects both `sugar.=>` and this PR's `~>` in some cases; should be easy to fix
* `alias2` and `aliassym` are placeholder names; I'll change these once we agree on final names
* I'd prefer the syntax: `alias x = foo.bar` instead of `const x = alias foo.bar`; this requires a minor parser change; it would simplify usage in general. The code in this PR wouldn't change much with this change, but user code would become simpler. If there's agreement, I'll switch to `alias x = foo.bar` and `proc bar(fun: alias)`

* compared to `sugar.=>`, the library code alias defined in this PR on top of `alias` are easier to use (doesn't suffer from https://github.com/nim-lang/Nim/issues/7435 as shown in unittests), and more efficient (no function call overheard that would prevent inlining)

* will hide behind a feature flag (eg: `{.experimental: aliassym.}`) once we agree on name

timotheecour commented:
just rebased (it was previously green but had started to bitrot) /cc @araq

timotheecour commented:
friendly ping /cc @araq

timotheecour commented:
rebased again against latest devel /cc @araq ; this PR (which garnered a lot of +1's) would fill an important gap in nim IMO;
I've been benefitting from it in my own code but it's unpleasant to have to keep rebasing a patched nim against latest devel :)

kristianmandrup commented:
Looks fantastic!! Please merge this ASAP, perhaps release it for `1.1.0`?

Araq commented:
This really needs a RFC process. Negative sides:

- I need to constantly think about this feature when adding new code to the compiler, comparable to ``skipTypes`` which is burdensome.
- It probably impacts the compiler's speed.

timotheecour commented:
@araq
> This really needs a RFC process.

=> https://github.com/nim-lang/RFCs/issues/232 ; i didn't think it would be a good idea to repeat in the RFC what was said in PR msg (DRY) but up for discussion in that RFC

> I need to constantly think about this feature when adding new code to the compiler, comparable to skipTypes which is burdensome.

correct, as is case with any new `TTypeKind`; note that the main added complexity from this PR is `tyAliasSym`, and appears 29 times; compared to 201+132 for `tyVar` + `tyLent`. I'm confident the feature is worth the cognitive cost, and will lead to further simplifications in stdlib + compiler; possibly even static and typedesc.

> It probably impacts the compiler's speed.

fair question given that `TTypeKind.high.ord` is now 64 (I don't see anything else adding a cost in this PR); however:
* best of 4 runs with `gtime -v ./koch boot -d:release` produced the same timings before and after this PR
(46s after PR vs 47 before PR, ie within measurement noise)
* likewise with `XDG_CONFIG_HOME= $nimb c -r testament/testament all`; (469s after PR on my local OSX; and 486s before PR, within measurement noise)
* deadcode `tyOpt` can be removed to bring this back to 63, although that probably won't impact compilation times based on above results

zah commented:
`tyFieldAccessor` is a type of symbol as well and can be merged into the new type kind. `tyUserTypeClassInst` is a type of alias and can be merged into `tyAlias`, but this requires some slightly more involved changes.

zah commented on tests/magics/tlambda.nim:
```diff
@@ -0,0 +1,351 @@
+
+proc testArrow() =
+  block: # 1 arg
+    const fun = a ~> (a, a)
```
Since this creates a template, would it suffer from the double evaluation problem of templates parameters? In other words, it I pass an expression that produces side effects (e.g. printing something to screen), it will be evaluated twice during the construction of the tuple.

timotheecour replied:
there are tests for that, it's side effect safe (no double evaluation):
```nim
block:
  const fun = a ~> (a, a)
  var count = 0
  proc identity(b: int): int = (count.inc; b)
  doAssert fun(identity(3)) == (3,3)
  doAssert count == 1
```
(`~>` is a pure library implementation on top of `alias`, see std/lambdas)

zah commented:
This PR demonstrates interesting capabilities. The fact that higher-order iterators can be compiled with it is particularly impressive. The downside is that some typing is lost in the process, but I think the current implementation could serve as a basis for a more complete one that:

1) Allows you to define templates and iterators taking a properly typed other templates and iterators.
2) Allows you to define generic types that are parametric over symbols (something that the examples here haven't tried to explore yet).

The PR in general can benefit from better naming for everything (both for the internal procs and for the user APIs).

timotheecour commented:
> The downside is that some typing is lost in the process ... current implementation could serve as a basis for a more complete one that

well it's still type-safe but yes, future PR's can introduce type constraints via `aliassym[T]`, eg: `aliassym[module]` or `aliassym[int|float]` or `aliassym[type]` or `aliassym[static]`  (exact syntax TBD)

note that with https://github.com/nim-lang/Nim/pull/12048 this should already work (and allow arbitrary constraints):
`proc fn(a: aliassym) {.enableif: a.sizeof > 8.}` ()


> Allows you to define generic types that are parametric over symbols (something that the examples here haven't tried to explore yet).

did you see `testAliasFields` in the test suite?
```nim
block:
  proc testAliasFields() =
    # this could be improved, by supporting:   `type B = object: fun1: aliassym fun2: aliassym`
    type A[T1, T2] = object
      fun1: T1
      fun2: T2
    proc initA[T1, T2](fun1: T1, fun2: T2): A[T1, T2] =
      A[T1, T2](fun1: alias2 fun1, fun2: alias2 fun2)
    const z = 123
    const a = initA(a~>a*10, alias2 z)
    doAssert a.fun1(3) == 30
    static: doAssert a.fun2 == 123
  testAliasFields()
```

> The PR in general can benefit from better naming for everything (both for the internal procs and for the user APIs).

yes, my plan is to make `alias` a keyword so that you'd be able to write `alias myecho = echo` instead of `const myecho = alias2 echo` and `bar(alias fun)` instead of `bar(alias2 fun)`; but this doesn't have to be in same PR

zah commented:
I'd like the following to work:

```nim
import
  lambdas

proc testAliasFields() =
  type A[Sym] = object
    x: int

  proc initA(Sym: aliassym, x: int): A[Sym] =
    A[Sym](x: x)
  
  var a = initA(a ~> a*10, 10)
  doAssert a.Sym(3) == 30
```

timotheecour commented:
* @zah me too; I've now implemented it, see testAliasGeneric and testAliasFields
* pushed another commit that shows that `alias` can completely replace `static[T]`

```nim
proc testStatic() =
  # aliassym can replace `static[T]`
  proc fn(a, b: aliassym, c: int): string =
    const a2 = a # sanity check to make sure it's static
    const b2 = b
    doAssert a is int
    doAssert b is seq[string]
    $(a, a2, b, b2, $type(a), $type(b), c)
  doAssert fn(lambdaStatic 12, lambdaStatic @["foo", "bar"], 13) == """(12, 12, @["foo", "bar"], @["foo", "bar"], "int", "seq[string]", 13)"""
```

* pushed another commit that shows that `alias` can also replace `typedesc[T]`
```nim
proc testTypedesc() =
  # aliassym can replace `typedesc[T]`
  proc fn(t1, t2, t3: aliassym): string =
    var a1: t1
    doAssert a1 is float32
    doAssert t1 is float32
    doAssert float32 is t1
    var a2: t2
    doAssert t2 is uint8
    result = $($t1, t1.default, $t2, $t3)
  type Foo[T] = object
  doAssert fn(lambdaType float32, lambdaType type(1u8+2u8), lambdaType Foo[int]) == """("float32", 0.0, "uint8", "Foo[system.int]")"""
```

zedeus commented on lib/pure/lambdas.nim:
```diff
@@ -80,6 +80,13 @@ template lambdaStatic*(a: untyped): untyped =
+  type T = typeof(block: (var a: t; a))
+  # type T = t
+  #   # this would work but would create a visible abstraction, eg:
+  #   # $ would render as: "T`gensym37245237"
```
Doesn't `{.dirty.}` fix that?

timotheecour replied:
this wouldn't work, it'd give `Error: redefinition of 'T'` error

zah commented:
Great, that's progress, but my example still fails here:

``` nim
  proc initA(Sym: aliassym, x: int): A[Sym] =
    # Error: type mismatch: got <A[template (a_fakegensym_1: untyped): untyped]> but expected 'A[aliasSym]'
    A[Sym](x: x) 
```

I can see that in your test you've worked-around the problem by using `type(result)(x: x)`, but that's a bit unfair. I think this PR represents a great enhancement of the language, but it would be better if we make it very solid and completely fleshed out before merging (potentially with the `alias` keyword and everything the community agrees on in the RFC).

Replacing `tyStatic` and `tyTypeDesc` with `tyAliasSym` is not a useful direction to pursue IMO. 

`tyStatic` represents the knowledge of the compiler that some expression evaluates to a known value at compile-time. A complete implementation of Nim would inject `tyStatic` automatically in all the right places during sem processing without relying on the lazily-triggered static evaluation that is used at the moment.

The problems with `tyStatic` stem from the need for delayed evaluation when the static expression contains unresolved generic paramenters. `tyAliasSym` cannot help with this problem.

The other applications of `tyStatic` are related to overload resolution where you use it to precisely specify what expression can be accepted. Now, that's the useful direction that this PR must pursue. It should be possible for us to hide many of the `tyAliasSym` usage behind more precise user-facing types:

```nim
type
  Foo1[Sym: template] = object
  Foo2[Sym: iterator(x: int): float] = object
  Foo3[Sym: aliassym[Callable[int, string]]] = object
```

Some nice outcomes of this has been demonstrated here:
https://github.com/nim-lang/RFCs/issues/212

Regarding naming, I think we can call `aliassym` just `symbol`. `tyTypeDesc` is a type of symbol and the arguments for not replacing it are weaker, but again, the user benefits from having easier to type user-facing types that may or may not be mapped internally to `tyAliasSym`.

For `alias` as a keyword similar to `const`, `var` or `let`, I think we might have enough rationale. When used with a symbol, it gets the treatment in this PR. When used over a run-time location, it tries to preserve the `tyLent`/`tyVar` properly of the location. It can perhaps error out if the location doesn't have this property (although C++ is accepting references to temporary objects through life extension and I've tended to rely on it).

zah commented:
Here is another test case that triggers an error:

```nim
import
  lambdas

const elementType = alias2 lambdas.elementType2

proc toSeq[T](a: T): auto =
  type T = elementType(a)
  result = newSeq[T]()
  for x in a: result.add x

iterator iota(n: int): auto =
  for i in 0..<n: yield i

iterator map[T, T2](a: T, fun: T2): auto =
  for ai in a: yield fun(ai)

echo toSeq(lambdaIter map(lambdaIter map(lambdaIter iota(3), a~>a*10), c~>c*10))
```

If we start specifying more precise types and introduce more rules in sigmatch, the code can look like this:

```nim

proc toSeq[T](a: symbol[iterator: T]): seq[T] =
  result = newSeq[T]()
  for x in a: result.add x

iterator iota(n: int): int =
  for i in 0..<n: yield i

iterator map[From, To](iterable: symbol[iterator: From],
                       operation: symbol[template (x: From): To]): To =
  for e in iterable:
    yield operation(e)

echo toSeq(map(map(iota(3), a~>a*10), c~>c*10))
```

The next step would be to introduce some concepts to relax the sigmatch requirements a bit:

```nim

proc toSeq[T](a: Iterable[T]): seq[T] =
  result = newSeq[T]()
  for x in a: result.add x

iterator iota(n: int): int =
  for i in 0..<n: yield i

iterator map[From, To](iterable: Iterable[From],
                       operation: Callable[From -> To]): To =
  for e in iterable:
    yield operation(e)

echo iota(3).map(a~>a*10).map(c~>c*10).toSeq
```

And there you go, the final code that should serve as a motivation/goal for this PR.

timotheecour commented:
@zah 
> Great, that's progress, but my example still fails here:

that's by design, aliases are resolved early, otherwise the alias abstraction would leak, making the feature a lot less useful.
```nim
  proc initA(Sym: aliassym, x: int): A[Sym] =
    # A[Sym](x: x) # this cannot work by design, `Sym` is resolved early.
    typeof(result)(x: x) # works
    A[alias2 Sym](x: x) # also works
```
> I can see that in your test you've worked-around the problem by using type(result)(x: x), but that's a bit unfair.

that was not a workaround; both `typeof(result)` and `A[alias2 Sym]` work as designed.

> Replacing tyStatic and tyTypeDesc with tyAliasSym is not a useful direction to pursue IMO.

I disagree but this can be discussed after this PR is merged, it's a not a blocker for this PR either way.

> A complete implementation of Nim would inject tyStatic automatically in all the right places during sem processing without relying on the lazily-triggered static evaluation that is used at the moment.

there are downsides to that, including code size; either way this requires an RFC and experiments.

> The problems with tyStatic stem from the need for delayed evaluation when the static expression contains unresolved generic paramenters. tyAliasSym cannot help with this problem.

that's not the only problems with `static`, far from it. Also, see this for example: https://github.com/nim-lang/Nim/issues/9679
it fails with static, and works with alias (see test case `test9679` I just added)

> It should be possible for us to hide many of the tyAliasSym usage behind more precise user-facing types:

I agree we need a way to restrict symbol kinds (and even beyond that), but this extension can be done in future work without breaking change; the syntax you're suggesting would be ambiguous, this already has a meaning before this PR:
```nim
proc fun[Sym: iterator(x: int): int](a: Sym) = discard
iterator myiter(a: int): int {.closure.} = yield 10
fun(myiter)
```

this other syntax can work though: `symbol[iterator: T]`, `symbol[module]`

> Regarding naming, I think we can call aliassym just symbol. 

`symbol` would be good but may cause name clashes (670 occurrences in nim repo, although a lot of these are in comments). I'm not married to `aliassym` but something unique/easy to search / not causing breaking changes would be nice to have. `symalias` is another good option. Also note that `aliassym` by design can also match against overloaded symbols.

timotheecour commented:
@zah 
> Here is another test case that triggers an error:
https://github.com/nim-lang/Nim/pull/11992#issuecomment-643213943

fixed it in latest commit, see tests `testArrowWrongSym2` + `testArrowWrongSym`

Varriount commented on lib/pure/lambdas.nim:
```diff
@@ -0,0 +1,111 @@
+#[
```
This module will need to be documented, along with all the callables it exposes.

timotheecour replied:
done, added doc comments + runnableExamples for everything

Varriount commented on lib/pure/lambdas.nim:
```diff
@@ -0,0 +1,111 @@
+
+import std/macros
+
+template lambdaIter*(a: untyped): untyped =
```
Is there anything that makes this useful only for iterators? I might be reading it incorrectly (I don't quite understand alias2), but this template seems to capture the passed expression and return a callable that invokes that expression.

Varriount commented on lib/pure/lambdas.nim:
```diff
@@ -0,0 +1,111 @@
+      body[i] = replaceSym(body[i], sym)
+  return body
+
+macro `~>`*(lhs, rhs: untyped): untyped =
```
The implementation of this could use some comments explaining what the goal of the code is, ie, "this goal of this macro is to generate a template that ... ")

timotheecour replied:
done, added doc comments + runnableExamples

planetis-m commented:
This looks like a new language construct. Maybe put it under an expiremental flag?

zah commented:
> that's by design, aliases are resolved early, otherwise the alias abstraction would leak, making the feature a lot less useful.

```nim
  proc initA(Sym: aliassym, x: int): A[Sym] =
    # A[Sym](x: x) # this cannot work by design, `Sym` is resolved early.
    typeof(result)(x: x) # works
    A[alias2 Sym](x: x) # also works
```

I'm not sure I understand the argument here. The expression `A[Sym]` should resolve to the same type, no matter how many times I use it within the body of the proc. Why should the return type expression be more special than any other expression?

timotheecour commented:
@b3liever
> This looks like a new language construct. Maybe put it under an expiremental flag?

done, and done in a way that can be localized via `{.push experimental:"alias".} ... {.pop.}` as opposed to "all or nothing"

Araq commented on compiler/sem.nim:
```diff
@@ -28,6 +28,9 @@ when not defined(leanCompiler):
 # implementation
 
+proc isMacroRealGeneric*(s: PSym): bool {.importc.}
+proc semAlias2(c: PContext, n: PNode): PNode {.exportc.}
```
We don't use ``importc`` and ``exportc`` to circumvent the cyclic import restrictions.

timotheecour replied:
we don't but we should, it's much better than relying on include (eg `include semtypes, semtempl, semgnrc, semstmts, semexprs`), and also better for incremental compilation (less tight coupling);

but for the sake of focusing on this PR I've addressed your comment and removed importc/exportc

timotheecour commented:
PTAL; rebased again and addressed latest comment https://github.com/nim-lang/Nim/pull/11992#discussion_r516051927

I'd really like this to be merged so it can garner more exposure (hidden behind `--experimental:alias`, which can be localized via push); further updates to this feature can then be done from `devel` branch instead of having to rebase + resolve conflicts each time.

timotheecour commented:
@araq , replying to your questions from https://forum.nim-lang.org/t/7020#44131 here to avoid hijacking the other thread:

> Then how come I understand how it works in Python but not really in your implementation?

python's underlying type system is much simpler (it being dynamic), and for example disallows calling functions without explicit parens, whereas in nim (and D), templates/macros can be called without parens and this is in fact an important feature to implement things like setters, getters; plenty of code relies on it (`currentSourcePath` etc). Furthermore python is dynamic and doesn't distinguish var/let/const so all you need is `print2 = print`, or `call(print, 1)`. Nim (like D) must distinguish things like passing by procvar vs passing by symbol, or parens-less template/macro calls.

> tyAliasSym is a feature on top of aliases, it seems

This was explained in the manual PR, but I think this shows you're still missing the main point of this PR, which is passing symbols to routines (1st class symbols). When you pass a symbol (eg an `skTemplate`) to a routine `proc fn(a: auto)`, parameter `a` must have a type in the routine declaration. That type is `tyAliasSym` and makes `fn` implicitly generic.

```nim
when true:
  import std/lambdas
  proc fn(a: auto): auto = $(a, a)
  template bar: untyped = (echo "in bar"; 1)
  echo fn(bar) # calls bar once; then passes 1 to fn (type is `tyInt`)
  echo fn(alias2 bar) # passes bar symbol to fn (type is `tyAliasSym`), which then calls it twice in `$(a, a)`
```

The parameter type in the routine declaration is aliassym (and only accepts explicit symbol aliases as arguments), but inside the routine body, the alias is a transparent abstraction and `type(a)` resolves to whatever the symbol resolves to:

```nim
when true:
  import std/lambdas
  proc fn1(a: int) = echo a.type
  proc fn2(a: aliassym) = echo a.type
  template bar: untyped = (echo "in bar"; 1)
  fn1(bar) # prints "in bar"; int (template called before being passed as an int)
  fn2(alias2 bar) # prints int (template passed as a symbol; a.type then resolves to int, not aliassym (transparent abstraction)
  # fn1(alias2 bar) # Error: type mismatch: got <aliasSym> ...
  # fn2(bar) # Error: type mismatch: got <int literal(1)> ...
```

Note that without `tyAliasSym`, it'd be impossible to declare the routine signature accepting symbol aliases (eg, what would it be for a routine accepting an skModule? or an skTemplate/skMacro like `echo` ? or an overloaded symbol like `$`?)

> Your "alias" design is biased: All it cares for is that examples A, B, C then "compile and work", but we also need to ensure that stuff that shouldn't compile does not suddenly compile...

that's not true, see `type mismatch` errors above. Aliases are always explicit and the design ensures there  are 0 ambiguities, including when dealing with templates that can be called without parenthesis.

> how it affects the compiler's error messages

sigmatch errors are explicit about what was expected vs given, as shown above

> there is considerable overlap with using const for proc aliases and with type for type aliases. If const doesn't work well enough, what would it take to make work well? Refining existing features is preferable over adding new ones.

This has been already answered in detail elsewhere. Existing features in language cannot alias reliably (neither via `const fun2 = fun` nor via `template fun2(a: varargs[untyped]): untyped` or `type Foo2 = Foo`) and do so with a non-transparent abstraction (unlike `alias`). More importantly, they're entirely incapable of the main feature in this PR, which is to allow passing symbols to routines.

This PR enables all the following features with a single concept:
* aliasing (no quirks/edge cases)
* lambdas (unlike sugar.=>, 0 cost and typing not required) eg `sort(myContainer, (x,y) ~> x.v-y.v)` in library code
* passing any symbol to any routine
* subsumes `static[T]` without any of its quirks/bugs
* subsumes `typedesc[T]` without any of its quirks/bugs.

If that's not an economy of means, I don't know what is.

Varriount commented:
To be fair, a symbolic lambda isn't the same as a closure. Symbols have to be computed at compile time. 

Regarding procvars, ideally a procedure like `proc foo(x: int, y = ""): int` could be converted to `proc foo(x: int): int` via an automatically generated thunk.

It would be wonderful if a procedure could just declare that it wants a callable that can accept and return a certain set of types, and the compiler could do the work of wrapping (or not wrapping) passed in values in procedures and/or closures.

Araq commented:
> This was explained in the manual PR, but I think this shows you're still missing the main point of this PR, which is passing symbols to routines (1st class symbols). When you pass a symbol (eg an skTemplate) to a routine proc fn(a: auto), parameter a must have a type in the routine declaration. That type is tyAliasSym and makes fn implicitly generic.

So just like ``auto`` and ``untyped``. Again, considerable overlap with existing mechanisms.

> Note that without tyAliasSym, it'd be impossible to declare the routine signature accepting symbol aliases (eg, what would it be for a routine accepting an skModule? or an skTemplate/skMacro like echo ? or an overloaded symbol like $?)

This is the core of the argument. And it's a big problem, you *underspecify* what the generic proc really needs. It's why Nim has a two-phase symbol lookup process for generics much like C++. With the work on concepts we're trying to get out of this mess, you simply need to describe precisely the generic constraints and then it's also clear that when you write ``a == b`` and ``a``'s type is "Comparable" you mean a's equality operator here and not "system.`==` for integers and strings and floats, packed into a SymChoice, to be resolved later".

You write ``proc ab(x: aliassym)`` and then you have to explain the constraints in the documentation, come on, that's simply not good enough anymore.

Araq commented:
> subsumes typedesc[T] without any of its quirks/bugs.

Ok, how do I write this proc with aliassym:

```nim
proc default[T](x: typedesc[T]): T
```

Note that ``default`` takes a type and returns a value (!) and ``T`` is a type variable that is inferred.

> If that's not an economy of means, I don't know what is.

First of all, we are talking about two different features with the same name, so that's a bit dishonest here. Secondly, "economy of means" is not convincing either, "I just replaced if/case/while/for by goto, economy of means!" I like the "Principle of Least Privilege" better.

timotheecour commented:
@Varriount
> To be fair, a symbolic lambda isn't the same as a closure. Symbols have to be computed at compile time.

not true:
```nim
when true:
  import std/lambdas
  proc call(a: aliassym): auto = a()
  var z = 0
  doAssert call(() ~> (z.inc; z*10)) == 10
  doAssert call(() ~> (z.inc; z*10)) == 20
```

@Araq 
> So just like auto and untyped. Again, considerable overlap with existing mechanisms.

`untyped` has no overlap with aliassym; `untyped` accepts an AST (no semcheck), whereas aliassym accepts a (semchecked) symbol alias.
`auto` matches all arguments with a type, so by definition it will also also match an alias; aliassym is implicitly generic because each distinct symbol passed to routine creates a different generic instantiation.

> This is the core of the argument. And it's a big problem, you underspecify what the generic proc really needs
> You write proc ab(x: aliassym) and then you have to explain the constraints in the documentation, come on, that's simply not good enough anymore.

As I've already explained earlier, you can optionally add arbitrary constraints **exactly as in C++ (`enable_if`), D (`if constraint`), and rust (`where`)** (https://blog.mgattozzi.dev/understanding-where-clauses/) via `enableif` (https://github.com/nim-lang/Nim/pull/12048).
I've rebased this PR against enableif, and you can try it out here: https://github.com/timotheecour/Nim/tree/pr_aliassym_enableif

this works:
```nim
when true:
  import std/lambdas
  import std/macros
  macro isTemplate(a: typed): bool = newLit a.symKind == nskTemplate
  proc bar(a: aliassym, b: auto): auto {.enableif: isTemplate(a).} = a(b); a(b)
  template f1(a: int) = discard
  proc f2(x: int) = discard
  bar(alias2 f1, 0)
  doAssert not compiles bar(alias2 f2, 0)

  proc bar2(a: aliassym, b: auto): auto {.enableif: compiles(a(b)).} = a(b); a(b)
  bar2(alias2 f1, 0)
  bar2(alias2 f2, 0)
  doAssert not compiles bar2(alias2 f2, 0.0)
```

you can refer to generic arguments as well, eg this also works:
```nim
when true:
  import std/lambdas
  import std/macros
  template returnSameType(a, T): bool = a(T.default) is T
  proc bar[T](a: aliassym, b: T): auto {.enableif: returnSameType(a, T).} = a(b)
  template f1(x: int): auto = x*x
  template f2(x: float): auto = x
  template f3(x: float): auto = cast[int](x)
  echo bar(alias2 f1, 0)
  echo bar(alias2 f2, 0.0)
  doAssert not compiles bar(alias2 f3, 0.0) # returnSameType would be false
```

Syntax sugar could in future be added in enableif to allow writing this:
```nim
# when a single argument is constrained:
proc bar(a: isTemplate, b: auto): auto = a(b); a(b)
# when a more complex expression is needed:
proc bar[T](a: returnSameType(a, T), b: T): auto = a(b)
```

> Ok, how do I write this proc [default] with aliassym:

```nim
import std/lambdas
proc default2Impl(a: aliassym): auto = (var ret: a; ret)
template default2(a): untyped = default2Impl(alias2 a)
doAssert float.default2 == 0.0
var z = 0
doAssert not compiles(z.default2)
```

or, if you want to constrain before instantiation, same answer as above:
```nim
import std/lambdas
proc default2Impl(a: aliassym): auto  {.enableif: a is type.} = (var ret: a; ret)
template default2(a): untyped = default2Impl(alias2 a)
```

This PR (symbol alias) and enabelif PR work together out of the box, even though they were written independently; those features compose well.

Araq commented:
```nim

proc default2Impl(a: aliassym): auto = (var ret: a; ret)
template default2(a): untyped = default2Impl(alias2 a)

```

But now the type signature became much less precise. We are not going to replace ``typedesc[T]`` anywhere with ``aliassym`` if it's less precise. aliassym doesn't "subsume typedesc[T] without any of its quirks/bugs". Likewise ``enableif`` is not the same as Rust's ``where`` clause. Generic code has to become more robust, Rust with its type checked generics is already there, C++ with its concepts is trying to get out of the duck-typing lazy generics too (much like Nim), D is still in denial that there is a problem. 

If you write more and more generic code (and that's what many Nim programmers want), you need to have more and more type-checking, I don't understand what's controversial about it. The list of programming language communities which have no plans of adding the alias/enableif features is quite long: C#, F#, Java, Swift, Go (they're discussing generics and in their proposal they type-check their generic proc bodies), Rust, ... Of course, maybe these languages are all misguided and would be better off with alias and enableif. But taken together with my other arguments it's yet another strong indicator.

Araq commented:
> untyped has no overlap with aliassym; untyped accepts an AST (no semcheck), whereas aliassym accepts a (semchecked) symbol alias.

Maybe ok, but ``typed{nkSym, nkOpenSymChoice, nkClosedSymChoice}`` comes close.

timotheecour commented:
> Maybe ok, but typed{nkSym, nkOpenSymChoice, nkClosedSymChoice} comes close.

`typed` has no relation with alias:
* `typed` can only be passed to a **template/macro** and pastes its semchecked AST, `alias` can be passed to **any routine** and passes the symbol

The fact that you can pass symbols to any routine (including proc's) is the most fundamental feature.

* a `typed` parameter in a macro is a NimNode, a aliassym param in a symbol, from which its type is derived, as usual:
```nim
when true:
  import macros
  import std/lambdas
  macro bar1(a: typed): untyped = newLit $a.type
  macro bar2(a: aliassym): untyped = newLit $($a.type, a(2))
  template fn1(b: int): untyped = b
  echo bar1(fn1) # NimNode
  echo bar2(alias2 fn1) # ("template (b: int): untyped", 2)
```
* likewise, typed param in a template also behaves differently from aliassym:

```nim
when true:
  import macros
  import std/lambdas
  template bar1(a: typed{nkSym, nkOpenSymChoice, nkClosedSymChoice}, b) = a(b)
  template bar2(a: typed, b) = a(b)

  template fn1(b: int) = discard
  template fn2(b = 0) = discard
  # bar1(fn1, 0) # Error: type mismatch: got <template (b: int), int literal(0)>
  bar2(fn1, 0)
  # bar2(fn2, 0) # Error: in expression 'discard (0)': identifier expected

  # with aliassym, you can pass to a proc (or any routine)
  proc bar3(a: aliassym, b: auto) = a(b)
  bar2(alias2 fn1, 0) # ok
  bar2(alias2 fn2, 0) # ok
```

> But now the type signature became much less precise
> you need to have more and more type-checking, I don't understand what's controversial about it

maybe you missed the line `or, if you want to constrain before instantiation, same answer as above: ...`, you can achieve exactly the same as typedesc[T], and more.

**alias+enableif indeed does subsume typedesc[T], static[T], concepts etc..., with a much simpler implementation, much less bugs, and much more flexible**. concepts and static are easily some of the most buggiest parts of nim, as evidenced in github issues.

alias+enableif just works, and (pending adding sugar to enableif) can also have the syntax `proc bar1(a: typedesc2[SomeInteger])`:

```nim
import std/[lambdas,macros]
template typedesc2[T](a): bool = a is type and a is T
template static2[T](a): bool = a is static and a is T and a isnot type
proc bar1(a: aliassym): auto {.enableif: typedesc2[SomeInteger](a).} = $a
proc bar2(a: aliassym): auto {.enableif: static2[SomeInteger](a).} = $a
when false: # or, with syntax sugar that can be added in future:
  proc bar1(a: typedesc2[SomeInteger]): auto = $a
  proc bar2(a: static2[SomeInteger]): auto = $a
doAssert bar1(alias2 int8) == "int8"
var z: int
doAssert not compiles bar1(alias2 z) # not a type
doAssert not compiles bar1(alias2 float) # not SomeInteger
const z1 = 1
const z2 = 1.0
let z3 = 1
doAssert bar2(alias2 z1) == "1"
doAssert not compiles bar2(alias2 z2) # not SomeInteger
doAssert not compiles bar2(alias2 int) # a skType, not a skConst
doAssert bar2(lambdaStatic 1) == "1" # 1 is not a symbol, but you can use `lambdaStatic` to wrap it into one
```

Araq commented:
> concepts and static are easily some of the most buggiest parts of nim, as evidenced in github issues.

I agree completely. But I see concepts as our ticket to get typechecked generics and I don't see how "enableif" can get us there.

> alias+enableif indeed does subsume typedesc[T], static[T], concepts etc..., with a much simpler implementation, much less bugs, and much more flexible.

Well, if that's true, that's really great, don't get me wrong. For now it doesn't even subsume ``typedesc[T]``. **How to infer the T?**

timotheecour commented:
> For now it doesn't even subsume typedesc[T]. How to infer the T?

here:
```nim
when true:
  import std/lambdas
  template typedesc2[T](a): bool = a is type and a is T

  # this currently works:
  proc bar1(T: aliassym): auto = T.default
  doAssert bar1(alias2 float) == 0.0
  proc bar2(T: aliassym): auto {.enableif: typedesc2[SomeInteger](T).} = T.default
  doAssert bar2(alias2 uint8) == 0'u8

  when false:
    # this hasn't yet been implemented but will work:
    proc bar3(T: aliassym): seq[T] = discard
    # pending future enableif syntax sugar:
    proc bar2(T: typedesc2[SomeInteger]): seq[T] = discard
```


> I see concepts as our ticket to get typechecked generics and I don't see how "enableif" can get us there.

`{.enableif.}` is much simpler and more powerful, it can express constraints that can't be expressed with concepts (`isIterable`, constraints involving more than 1 arguments, constraints involving `untyped`, etc) without all the bugs.

**just look at how much easier it is to write a correct+efficient `toSeq` with this PR**.
As the author of https://github.com/nim-lang/Nim/pull/8711, I can tell you how refreshing it is to be able to use `proc` instead of `template` and not have to worry about `evalOnceAs` (or all the gymnastics in `toSeq`, `toSeq1`, `toSeq2`).

Also note that sequtils.toSeq is still incorrect for openArray (multiple evaluation bug, and it's really hard if not impossible to fix once you understand the problem), whereas with this PR, it just works:

```nim
import std/lambdas
template isIterable(a): bool = typeof(block: (for ai in a: ai)) is type
proc toSeqImpl(a: auto): seq {.enableif: isIterable(a).} =
  type T = typeof(block: (for ai in a: ai))
  when compiles(a.len):
    result = newSeq[T](a.len)
    var i=0
    for ai in a: (result[i] = ai; i.inc)
  else:
    result = newSeq[T]()
    for ai in a: result.add ai

template toSeq(a: untyped): untyped = toSeqImpl(lambdaIter a) # sugar to avoid calling code having to write `lambdaIter`

doAssert toSeq(10..13) == @[10,11,12,13]
iterator iota(n: int): int = (for i in 0..<n: yield i)
doAssert toSeq(iota(3)) == @[0,1,2]
doAssert @[10,11].toSeq == @[10,11]
doAssert [10,11].toSeq == @[10,11]
doAssert not compiles toSeq 10

# works with openArray, safe wrt multiple evaluation bug, unlike sequtils.toSeq!
from sequtils import nil
proc fn3(a: openArray[int]) =
  template fn(x: int): untyped = (echo "in fn"; x)
  echo "this PR toSeq: no bug"
  doAssert toSeq(toOpenArray(a,0,fn(1))) == @[10,11]
  echo "sequtils.toSeq: multiple evaluation bug and hard/impossible to fix"
  doAssert sequtils.toSeq(toOpenArray(a,0,fn(1))) == @[10,11]
fn3 @[10,11]
```

prints:
```
this PR toSeq: no bug
in fn
sequtils.toSeq: multiple evaluation bug and hard/impossible to fix
in fn
in fn
in fn
in fn
in fn
in fn
```

Araq commented:
As long as your return type remains ``auto`` you don't have a solution that is satisfying. Also, how can you "enableif" constraints for the return type? Remember that the documentation generator cares about proc headers, not about proc bodies.

SixteNim commented:
``aliassym`` simulates a ``#define name (....)`` well known from C, but locally scoped instead of C's rather undefined (global) scopes. If the ``(...)`` stands for a name itself, alpha-conversion will happen - the most basic (trivial) substitution. Therefore, a proc that annotates a   `` a :  aliassym b`` can be read as a substitution  a for b , where the proc body uses an "a" that stands for a "b" and the "b" itself stands for a witness, for both  a  prototypic implementation and  a pattern for other substitutions as well. Logically, ``aliassym`` is sound - as long as a witness is presented. But what about substitutions? Substitutions should be made possible with an attribute, e.g. ``{.  witness 'name' .}`` where "name" stands for the name of the witness. The attribute is sufficient enough to get a sound ``aliassym`` , no``enableif`` etc. needed. As long as scope borders are respected, the implementer of a particular module will have full control about the module's intended behaviour. This is important for safety reasons. If scope borders are not respected, the compiler should give a warning (at least).  
``aliassym`` is crucial at every place  where structural substitution is needed (iterators) ore desired (lambdas to be inlined). It would  improve Nim's expressiveness, safety and speed at the same time. Why don't other languages (except D)  have this feature? Well, they are not flexible enough at the AST level.

mwerezak commented:
Notwithstanding the technical discussion in the PR, having first class symbols and composable iterators would be a huge boost to expressiveness of the language.

mwerezak commented:
rip

ringabout commented:
Although this PR is incredible. It has been a struggle to maintain the existing features of the Nim language. What really is needed for now is to refactor existing features and make it accessible among backends. Anyway the stale label means eveyone could pick it up and complete it.
